lib/src/extension.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'default_extension_map.dart';
+
+/// Add an override for common extensions since different extensions may map
+/// to the same MIME type.
+final Map<String, String> _preferredExtensionsMap = <String, String>{
+  'application/vnd.ms-excel': 'xls',
+  'application/vnd.ms-powerpoint': 'ppt',
+  'image/jpeg': 'jpg',
+  'image/tiff': 'tif',
+  'image/svg+xml': 'svg',
+  'text/calendar': 'ics',
+  'text/javascript': 'js',
+  'text/plain': 'txt',
+  'text/sgml': 'sgml',
+  'text/x-pascal': 'pas',
+  'video/mp4': 'mp4',
+  'video/mpeg': 'mpg',
+  'video/quicktime': 'mov',
+  'video/x-matroska': 'mkv',
+};
+
+/// Lookup file extension for a given MIME type.
+///
+/// If there are multiple extensions for [mimeType], return preferred extension
+/// if defined, or the first occurrence in the map.
+/// If no extension is found, `null` is returned.
+String? extensionFromMimeOrNull(String mimeType) {
+  final mimeTypeLC = mimeType.toLowerCase();
+  if (_preferredExtensionsMap.containsKey(mimeTypeLC)) {
+    return _preferredExtensionsMap[mimeTypeLC]!;
+  }
+  return defaultExtensionMap.entries
+      .firstWhereOrNull((entry) => entry.value == mimeTypeLC)
+      ?.key;
+}
+
+String extensionFromMime(String mimeType, {String? orElse}) =>
+    extensionFromMimeOrNull(mimeType) ?? orElse ?? mimeType;
+
+/// Allow for a user-specified MIME type-extension mapping that overrides the
+/// default.
+void addMimeType(String mimeType, String extension) {
```

Comment:
I'd be more comfortable with this PR if it didn't include this API.

Why do we need this one?

Reply:
hi @natebosch , thanks for having a look

this library does not know all possible file formats and I think it's neither realistic nor a goal of this lib to achieve that. The library also seems to be seeded with quite a random list of formats. So it can be quite a frequent situation in which a user of this library needs a format added. A pull request could be made, but the release cycle is not that fast and it's also a bit of overhead for maintainers.

That's why I think it may be useful to have this method so users don't immediately get stuck.

Similarly, this library already provides `addExtension`. 

The 2nd usecase is one in which the user of the lib has a different preference than the one that is included. For example, if one prefers 'text/plain' to map to 'conf' as it currently does, one can still configure that old behavior.

Reply:
Sorry for the slow response.

I'm worried about having an `addMimeType` that only impacts [extensionFromMime](https://pub.dev/documentation/mime/latest/mime/extensionFromMime.html) and doesn't impact [lookupMimeType](https://pub.dev/documentation/mime/latest/mime/lookupMimeType.html).

We could either make this signature `addMimeType(String mimeType, String extension, {List<int>? headerBytes})` and try to make this work with `lookupMimeType`, or we could consider an alternative name like `addExtensionPreference.

@devoncarew @lrhn WDYT?

Reply:
Generally, I worry about global mutable state.
This is a kind of registry, where any code can add its own preferences, but they then also affect *everybody else*.

I'd prefer a structure where you have a `Mime` *object* instead of global state and functions. Then you can get your own object, modify it, and pass it around to your own code, without affecting other people's `Mime` objects.
That's a larger change, though.

Reply:
I don't think we'll be able to do any refactoring of existing usage.

We could add a `MimeRegistry` class which can be overridden like this. The top-level APIs would continue to use the global map, the `MimeRegistry` would fall back on the global map, but it could also store local overrides. @lrhn WDYT about something like this?

Reply:
@lrhn - Should I work on a PR with an implementation of a `MimeRegistry`, or do you think we can get by with the top level approach?

Reply:
> What I'd do, if I had all the time in the world

What do you think we should do given that we have less time than that?

Reply:
The same thing! The worry about lack of time for something else afterwards :grin: 

But the concrete problem we're trying to solve here seems to be solved without the `addMimeType`, so I'd remove that, not add configurability in this PR, and just give access to the default extensions.

Reply:
removed `addMimeType`